### PR TITLE
Bidirectional GRU and LSTM C++ API forward fix

### PR DESCRIPTION
**Pull request by pbelevich**
Fixing #17998

### CODE REVIEW

**Comment by zou3519** [comment]
You've unintentionally updated the submodule "third_party/fbgemm". I think you can `cd` into it and then check out the version that is supposed to be there.

File: torch/csrc/api/src/nn/modules/rnn.cpp [hunk 1]
```diff
@@ -167,8 +169,8 @@ std::vector<Tensor> RNNImplBase<Derived>::flat_weights() const {
   // (w_ih, w_hh, b_ih, b_hh), repeated for each layer (next to each other).
   std::vector<Tensor> flat;
   const auto num_directions = options.bidirectional_ ? 2 : 1;
-  for (int64_t layer = 0; layer < options.layers_; layer++) {
-    for (auto direction = 0; direction < num_directions; direction++) {
+  for (auto direction = 0; direction < num_directions; direction++) {
+    for (int64_t layer = 0; layer < options.layers_; layer++) {
```
Comment (zou3519) [inline torch/csrc/api/src/nn/modules/rnn.cpp]: From a quick glance at the code, this seems wrong to me. `flat_weights` seems to be used in the same way as the Python counterpart, as [an argument to `_cudnn_rnn_flatten_weight`](https://github.com/pytorch/pytorch/blob/8399197df67f90d7e5cb52d735eb55c803664b35/torch/csrc/api/src/nn/modules/rnn.cpp#L119-L137). In that case, it should have the same order as the parameters in the python `flat_weights`, where the parameters are ordered as [[l0 l0_reversed l1 l1_reversed ...]](https://github.com/pytorch/pytorch/blob/8399197df67f90d7e5cb52d735eb55c803664b35/torch/nn/modules/rnn.py#L65-L86).

Can you add a test for CUDA LSTM forward?

File: test/cpp/api/rnn.cpp [hunk 2]
```diff
@@ -247,3 +247,80 @@ TEST_F(RNNTest, BidirectionalFlattenParameters) {
...
+              torch::tensor({5, 4, 3, 2, 1}, options).reshape({5, 1, 1});
+
+  LSTM bi_lstm {LSTMOptions(1, 1)
+                  .layers(1).batch_first(false).bidirectional(true)};
+  LSTM reverse_lstm {LSTMOptions(1, 1)
+                  .layers(1).batch_first(false).bidirectional(false)};
+
+  // Now make sure the weights of the reverse lstm layer match
+  // ones of the (reversed) bidirectional's:
+  reverse_lstm->w_ih[0].set_requires_grad(false).copy_(bi_lstm->w_ih[1]);
```
Comment (zou3519) [inline test/cpp/api/rnn.cpp]: nit: This is fine in test code but a user wouldn't write something like this (because this makes the parameters of the model un-trainable). Something like the following should work better:
```
{
  NoGradGuard guard;
  reverse_lstm->w_ih[0].set_requires_grad(false).copy_(bi_lstm->w_ih[1]);
  reverse_lstm->w_hh[0].set_requires_grad(false).copy_(bi_lstm->w_hh[1]);
  reverse_lstm->b_ih[0].set_requires_grad(false).copy_(bi_lstm->b_ih[1]);
  reverse_lstm->b_hh[0].set_requires_grad(false).copy_(bi_lstm->b_hh[1]);
}
```

File: torch/csrc/api/src/nn/modules/rnn.cpp (hunk 1 above)
Comment (zou3519) [reply torch/csrc/api/src/nn/modules/rnn.cpp]: I noticed that the tests are only testing for `layer=1` so swapping these for loops shouldn't have an effect

Comment (pbelevich) [reply torch/csrc/api/src/nn/modules/rnn.cpp]: I swapped these for loops to make it compatible with [existing RNN.cpp:gather_params() method](https://github.com/pytorch/pytorch/blob/8399197df67f90d7e5cb52d735eb55c803664b35/aten/src/ATen/native/RNN.cpp#L166:L181)

But I will write a test for CUDA LSTM forward to double check

File: test/cpp/api/rnn.cpp (hunk 2 above)
Comment (yf225) [reply test/cpp/api/rnn.cpp]: I would also recommend using
```cpp
reverse_lstm->w_ih[0].copy_(bi_lstm->w_ih[1])
```
instead of 
```cpp
reverse_lstm->w_ih[0].set_requires_grad(false).copy_(bi_lstm->w_ih[1])
```
because `NoGradGuard guard` would already disable gradient tracking of `copy_`.

File: torch/csrc/api/src/nn/modules/rnn.cpp (hunk 1 above)
Comment (zou3519) [reply torch/csrc/api/src/nn/modules/rnn.cpp]: Discussed offline, but leaving it here before I forget:
```
In [4]: m = torch.nn.LSTM(input_size=5, hidden_size=7, num_layers=3, bidirectional=True)
In [5]: m._get_flat_weights_names()
Out[5]:
['weight_ih_l0',
 'weight_hh_l0',
 'bias_ih_l0',
 'bias_hh_l0',
 'weight_ih_l0_reverse',
 'weight_hh_l0_reverse',
 'bias_ih_l0_reverse',
 'bias_hh_l0_reverse',
 'weight_ih_l1',
 'weight_hh_l1',
 'bias_ih_l1',
 'bias_hh_l1',
 'weight_ih_l1_reverse',
 'weight_hh_l1_reverse',
 'bias_ih_l1_reverse',
 'bias_hh_l1_reverse',
 'weight_ih_l2',
 'weight_hh_l2',
 'bias_ih_l2',
 'bias_hh_l2',
 'weight_ih_l2_reverse',
 'weight_hh_l2_reverse',
 'bias_ih_l2_reverse',
 'bias_hh_l2_reverse']
```
The python weights appear to be in that order, however, it's possible that they're completely named incorrectly

**Comment by zou3519** [comment]
Tests aren't running, this might need a rebase

File: test/cpp/api/rnn.cpp [hunk 3]
```diff
@@ -247,3 +248,250 @@ TEST_F(RNNTest, BidirectionalFlattenParameters) {
...
+// as regular forward of unidirectional GRU
+void BidirectionalGRUReverseForward(bool cuda) {
+  auto options = torch::TensorOptions()
+                        .dtype(torch::kFloat32).requires_grad(false);
+  auto input =
+              torch::tensor({1, 2, 3, 4, 5}, options).reshape({5, 1, 1});
+  auto input_reversed =
+              torch::tensor({5, 4, 3, 2, 1}, options).reshape({5, 1, 1});
+
+  GRU bi_grus {GRUOptions(1, 1)
```
Comment (zou3519) [inline test/cpp/api/rnn.cpp]: nit: You could also just create the GRUOptions once and then pass to both GRUs.

File: test/cpp/api/rnn.cpp [hunk 4]
```diff
@@ -247,3 +248,250 @@ TEST_F(RNNTest, BidirectionalFlattenParameters) {
...
+                  .layers(1).batch_first(false).bidirectional(true)};
+  GRU reverse_gru {GRUOptions(1, 1)
+                  .layers(1).batch_first(false).bidirectional(false)};
+
+  if (cuda) {
+    bi_grus->to(torch::kCUDA);
+    reverse_gru->to(torch::kCUDA);
+    input = input.to(torch::kCUDA);
```
Comment (zou3519) [inline test/cpp/api/rnn.cpp]: It's more efficient to create what you can on the GPU instead of moving things. `input` and `input_reversed` can be created on the GPU:
```
  auto options = torch::TensorOptions()
                        .dtype(torch::kFloat32).requires_grad(false).device(cuda ? kCUDA : kCPU);
```

**Review by zou3519 — APPROVED** [review]
lgtm, thanks! Some minor comments on making the tests nicer.
